File: web/frame.js
// Frame contract (spec §14) shared by the teaching toy model and the live Go
// poller. The renderer consumes this shape and never inspects `mode`. The Go
// type in internal/frame/frame.go must stay byte-for-byte compatible with this.
//
// @typedef {Object} Edge   { rate:number, unit:string, split?:{checkpointer,bgwriter,backend}, meanFsyncMs?:number, unavailable?:boolean }
// @typedef {Object} Node   { value:number, se:number, confident:boolean, estimated?:boolean, unavailable?:boolean }
// @typedef {Object} Frame  { ts, mode, edges:{eD,eA,eB,eC1,eC2}, nodes:{BM,WI,WW,EV}, levels:{dirtyFraction,cacheHit,dirtyEstimated?}, bottleneck:string|null, warmingUp:boolean }

// Canned fixtures used for step-1 frontend development and as a contract lock
// (render against these with no toy model and no DB).
const FIXTURES = {
  warmingUp: {
    ts: 0, mode: "live",
    edges: {
      eD: { rate: 0, unit: "records/s" }, eA: { rate: 0, unit: "blocks/s" },
      eB: { rate: 0, unit: "blocks/s", split: { checkpointer: 0, bgwriter: 0, backend: 0 } },
      eC1: { rate: 0, unit: "bytes/s" }, eC2: { rate: 0, unit: "syncs/s", meanFsyncMs: 0 },
    },
    nodes: {
      BM: { value: 0, se: 0, confident: false }, WI: { value: 0, se: 0, confident: false },
      WW: { value: 0, se: 0, confident: false }, EV: { value: 0, se: 0, confident: false, estimated: true },
    },
    levels: { dirtyFraction: 0, cacheHit: 0 }, bottleneck: null, warmingUp: true,
  },
  evRed: {
    ts: 0, mode: "live",
    edges: {
      eD: { rate: 5200, unit: "records/s" }, eA: { rate: 9100, unit: "blocks/s" },
      eB: { rate: 3400, unit: "blocks/s", split: { checkpointer: 200, bgwriter: 150, backend: 3050 } },
      eC1: { rate: 410000, unit: "bytes/s" }, eC2: { rate: 60, unit: "syncs/s", meanFsyncMs: 2.1 },
    },
    nodes: {
      BM: { value: 0.31, se: 0.03, confident: true }, WI: { value: 0.22, se: 0.03, confident: true },
      WW: { value: 0.28, se: 0.03, confident: true }, EV: { value: 0.82, se: 0.04, confident: true, estimated: true },
    },
    levels: { dirtyFraction: 0.78, cacheHit: 0.74 }, bottleneck: "EV", warmingUp: false,
  },
  allGreen: {
    ts: 0, mode: "live",
    edges: {
      eD: { rate: 1200, unit: "records/s" }, eA: { rate: 300, unit: "blocks/s" },
      eB: { rate: 800, unit: "blocks/s", split: { checkpointer: 700, bgwriter: 90, backend: 10 } },
      eC1: { rate: 120000, unit: "bytes/s" }, eC2: { rate: 30, unit: "syncs/s", meanFsyncMs: 0.9 },
    },
    nodes: {
      BM: { value: 0.08, se: 0.02, confident: true }, WI: { value: 0.12, se: 0.02, confident: true },
      WW: { value: 0.15, se: 0.02, confident: true }, EV: { value: 0.02, se: 0.01, confident: true, estimated: true },
    },
    levels: { dirtyFraction: 0.22, cacheHit: 0.99 }, bottleneck: null, warmingUp: false,
  },
};

File: internal/frame/frame.go
// Package frame defines the wire contract (spec §14) shared by both the
// teaching toy model (client-side JS) and the live Go poller. The frontend
// renders whatever frame arrives and never inspects which source produced it.
package frame

// Edge is one of the five animated edges. Rate drives animation speed
// (throughput); it is never colored by severity.
type Edge struct {
	Rate float64 `json:"rate"`
	Unit string  `json:"unit"`
	// Split is populated only for eB (flush), keeping the
	// checkpointer/bgwriter/backend breakdown. Nil for other edges.
	Split *FlushSplit `json:"split,omitempty"`
	// MeanFsyncMs is populated only for eC2 (WAL fsync). Nil otherwise.
	MeanFsyncMs *float64 `json:"meanFsyncMs,omitempty"`
	// Unavailable marks an edge whose source could not be read (missing
	// view/column/permission). The frontend renders it greyed-out (spec §16).
	Unavailable bool `json:"unavailable,omitempty"`
}

// FlushSplit breaks eB throughput down by writer.
type FlushSplit struct {
	Checkpointer float64 `json:"checkpointer"`
	Bgwriter     float64 `json:"bgwriter"`
	Backend      float64 `json:"backend"`
}

// Node is a contention node. Value (0..1) drives color (spec §12); it is
// independent of edge speed (spec §2.4).
type Node struct {
	Value     float64 `json:"value"`
	SE        float64 `json:"se"`
	Confident bool    `json:"confident"`
	// Estimated is set true only for EV, which is a composite signal (spec §9.3).
	Estimated bool `json:"estimated,omitempty"`
	// Unavailable marks a node whose source could not be sampled (spec §16).
	Unavailable bool `json:"unavailable,omitempty"`
}

// Levels carries the scalar readouts.
type Levels struct {
	DirtyFraction float64 `json:"dirtyFraction"`
	CacheHit      float64 `json:"cacheHit"`
	// DirtyEstimated marks dirtyFraction as derived rather than measured from
	// pg_buffercache (spec §16).
	DirtyEstimated bool `json:"dirtyEstimated,omitempty"`
}

// Frame is the per-tick payload (spec §14). Both modes emit this exact shape.
type Frame struct {
	TS         int64           `json:"ts"`
	Mode       string          `json:"mode"` // "teaching" | "live"
	Edges      map[string]Edge `json:"edges"`
	Nodes      map[string]Node `json:"nodes"`
	Levels     Levels          `json:"levels"`
	Bottleneck *string         `json:"bottleneck"` // "EV"|"WW"|"WI"|"BM"|null
	WarmingUp  bool            `json:"warmingUp"`
}

// Edge keys.
const (
	EdgeED  = "eD"  // backends -> shared buffers (modify/dirty)
	EdgeEA  = "eA"  // data files -> shared buffers (read-in on miss)
	EdgeEB  = "eB"  // shared buffers -> data files (flush)
	EdgeEC1 = "eC1" // backends -> WAL buffers (WAL insert)
	EdgeEC2 = "eC2" // WAL buffers -> WAL segments (fsync)
)

// Node keys.
const (
	NodeBM = "BM" // BufferMapping
	NodeWI = "WI" // WALInsert
	NodeWW = "WW" // WALWrite / fsync
	NodeEV = "EV" // backend dirty-eviction (composite)
)
